### PR TITLE
Clean up error handling for state-setting

### PR DESCRIPTION
**Pull request by jlowin**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] add a changelog entry in the `changes/` directory (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
Mirrors https://github.com/PrefectHQ/cloud/pull/2364 by raising errors more aggressively when states are set.

### CODE REVIEW

File: server/Changelog.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 ### Enhancements
 
 - new cli command--`prefect-server dev build` will build prefect server images from source locally - [#2227](https://github.com/PrefectHQ/prefect/pull/2227)
+- Failing to set a state raises errors more aggressively - [#2708](https://github.com/PrefectHQ/prefect/pull/2708)
```
Comment (joshmeek) [inline server/Changelog.md:12]: The way we update the changelog has changed. Use the process outlined here https://github.com/PrefectHQ/prefect/tree/master/changes

Comment (jlowin) [reply server/Changelog.md:12]: There's no new process for server?

Comment (joshmeek) [reply server/Changelog.md:12]: What we have been doing is moving the `server` section of the changelog up to the main changelog

Comment (jlowin) [reply server/Changelog.md:12]: Ah sorry

**Pull request merged**